MTB phylogenetics tutorial (#3220)
galaxyproject/training-material: topics/phylogenetics/tutorials/MTB phylogeny/tutorial.md

>
> 1. {% tool [Phyogenetic reconstruction with RAxML](toolshed.g2.bx.psu.edu/repos/iuc/raxml/raxml/8.2.4+galaxy2) %} with the following parameters:
>    - {% icon param-file %} *"Source file with aligned sequences"*: `output` (Input dataset)
>    - *"Model type"*: `Nucleotide`
>    - *"RAxML options to use"*: `Required options only`
>
> 2. The RAxML output we are interested in is the "Best-scoring ML tree". Select it in you Galaxy history and take a look at it with the different visualization options offered by Galaxy.
>
{: .hands_on}

# Visualize and manipulate the tree
Phylogenetic trees are great tools because they are at the same time quantitative (we can do calculations on branch lengths, estimate uncertainty of a tree topology etc.) and visually appealing, allowing to actually "see" biologically interesting patterns. Often this requires some tweaking of the tree, for example by coloring parts of the tree according to some background information we have about the samples. Here, to obtain an overview of the different strains present in the country (given the limitations of the sampling), our goal is to create a rooted tree in which colors indicate different MTB lineages.

> We will use R to plot and manipulate the phylogeny obtained from RAxML. The code to produce the figures is shown in the boxes below. You can execute it by starting RStudio within Galaxy, as explained [here]({% link topics/galaxy-interface/tutorials/rstudio/tutorial.md %}). This is not required to finish this tutorial, but if you have used R before, it might be worthwile to go through the code, modify it, and explore the numerous phylogenetics packages and functions in R.

Reviewer: ```suggestion
We will use R to plot and manipulate the phylogeny obtained from RAxML. The code to produce the figures is shown in the boxes below. You can execute it by starting RStudio within Galaxy. This is not required to finish this tutorial, but if you have used R before, it might be worthwile to go through the code, modify it, and explore the numerous phylogenetics packages and functions in R.

> ### {% icon hands_on %} Hands-on: Starting RStudio
>
> 1. Note the history IDs of the "Best-scoring ML tree"
>
> 2. {% tool [RStudio](interactive_tool_rstudio) %} in Galaxy provides some special functions such as `gx_get` to import and export files from your history.
>
>    {% snippet faqs/galaxy/interactive_tools_rstudio_launch.md %}
>
> 3. Copy the files we need into our workspace so we can see them in the Files pane.
>
>    ```R
>    treefile <- gx_get(1)  # Use the correct ID for your differentially expressed results dataset.
>    ```
>
{: .hands_on}
```

Reviewer: Sorry @cstritt when I suggested using `{% link` I wasn't reading closely enough, here's a more GTN-standard way to tell users to launch RStudio

Author: This looks good, thanks!

"ERR181435.fastq.vcf" = "Sensitive",
"ERR313115.fastq.vcf" = "Sensitive",
"ERR551620.fastq.vcf" = "MDR",
"ERR1203059.fastq.vcf" = "Sensitive",
"ERR2659153.fastq.vcf" = "Sensitive",
"ERR2704678.fastq.vcf" = "Sensitive",
"ERR2704679.fastq.vcf" = "Sensitive",
"ERR2704687.fastq.vcf" = "Sensitive",
"ERR5987300.fastq.vcf" = "PreXDR",
"ERR5987352.fastq.vcf" = "PreMDR",
"ERR6362078.fastq.gz.vcf" = "MDR",
"ERR6362138.fastq.gz.vcf" = "MDR",
"ERR6362139.fastq.vcf" = "PreMDR",
"ERR6362156.fastq.gz.vcf" = "PreXDR",
"ERR6362253.fastq.gz.vcf" = "MDR",
"ERR6362333.fastq.gz.vcf" = "PreXDR",
"ERR6362484.fastq.vcf" = "PreMDR",
"ERR6362653.fastq.gz.vcf" = "MDR",
"SRR998584.fastq.vcf" = "Sensitive",
"SRR13046689.fastq.vcf" = "Other"
)

tree_dr <- tree_rooted
tree_dr$tip.label <- as.character(mtbc_dr[tree_rooted$tip.label])

color_code_dr = c(
  Sensitive = "#ff00ff",
  PreXDR = "#0000ff",
  PreMDR = "#a000cc",
  MDR = "#ff0000",
  Other = "#663200")

pal_dr <- as.character(color_code_dr[tree_dr$tip.label])

par(mfrow = c(1, 2))
plot(tree_rooted,cex = 0.7, root.edge = T)
plot(tree_dr,cex = 0.8, tip.color = pal_dr, root.edge = TRUE)

```
<img src="./images/tree_rooted_dr.svg" width="600"/>

> ### Question

Reviewer: ```suggestion
```

---
layout: tutorial_hands_on
title: Tree thinking for tuberculosis evolution and epidemiology
zenodo_link: 'https://zenodo.org/record/6010176'
tags:
  - prokaryotes
  - phylogenetics
  - evolution
level: Beginner
objectives:
- Understand the basic concepts behind phylogenetic trees
- Be able to read and interrogate a phylogeny encountered in the literature
time_estimation: 1H
contributors:
- Christoph Stritt
- Daniela Brites

> - Galo A. Goig

Reviewer: these names should be references to the CONTRIBUTORS.yaml file, you'll need to add them in there.

Author: names were added

<img src="./images/mtbc_1strainPerLineage.nwk.COMB.svg" alt="drawing" width="800"/>
<figcaption align = "left">Fig.1 - Rooted (A) and unrooted (B) phylogeny of the MTBC. Some basic tree vocabulary is shown in red.</figcaption>

## Branch lengths
Besides relatedness and direction, a third important piece of information contained in a phylogeny is the branch length. When a phylogeny was estimated from DNA or protein sequences, branch length usually reflects the evolutionary distance between nodes in the tree. This information can be used to translate distance in terms of expected nucleotide changes into years, and thus to connect evolutionary change to historical events. Molecular dating is not covered in this tutorial, but if you should be interested in this topic, here is a [recent study](https://doi.org/10.1371/journal.ppat.1008067) showing the promises and problems of molecular dating with MTB.

> As branch lengths reflect evolutionary distances, they can also be used to identify transmission clusters and outbreaks. Below is a (rooted) tree of the Central Asian Clade (CAC), which is part of lineage 2 [(Eldholm et al. 2016)](https://doi.org/10.1073/pnas.1611283113). The orange color highlights the Afghan strain family within the CAC. At the bottom of the tree, note the clade with short branch lengths. This is how one would expect an outbreak to look in a phylogenetic tree: a set of strains clustering together and separated by extremely short branches, reflecting their almost identical genomes.

Reviewer: For all of these papers you reference, they could be better done like this:

```suggestion
As branch lengths reflect evolutionary distances, they can also be used to identify transmission clusters and outbreaks. Below is a (rooted) tree of the Central Asian Clade (CAC), which is part of lineage 2 ({% cite eldholm2016 %}). The orange color highlights the Afghan strain family within the CAC. At the bottom of the tree, note the clade with short branch lengths. This is how one would expect an outbreak to look in a phylogenetic tree: a set of strains clustering together and separated by extremely short branches, reflecting their almost identical genomes.
```

and then use a service like doi2bib.org to use convert the reference into a bibtex format and use in the `tutorial.bib` file

Author: This was changed. I couldn't preview the citations, so I don't know how they will be formatted.

<img src="./images/eldholm2016_tree.png" alt="drawing" width="200"/>
<figcaption align = "left">Fig.2 - Phylogeny of the central Asian clade, including the Oslo outbreak. Modified Eldholm et al. 2016.</figcaption>

> <br/><br/>

Author: I just removed the alt text, I hope that's fine

> <br/><br/>

Reviewer: The alt text is fine! But there's a couple different things together here in this suggestion:

- Please use `![alt-text](image-path "Figcaption text")` instead of manually doing `<img>` and `<figcaption>`
- Alt text is mandatory. And in this case it was good! 
- And notice that I've removed `Fig.2`, because when you use the markdown syntax for referencing images (with our addition of the ` "text"` as part of the path for the figure caption, the figures are numbered for you automatically.

> <br/><br/>

Author: Ok, I've changed the figure format now. Thanks for the details!

>
> Phylogenetics with MTB has some particularities not encountered with other organisms. Here are four points to keep in mind when doing phylogenetics with this organism.
>
> a) There seems to be **no horizontal gene transfer** (HGT) in the MTBC. HGT complicates phylogenetic inference in many bacteria because a piece of DNA introduced by HGT has a different history and thus phylogeny than genes not affected by HGT.
>
> b) There is **little genetic diversity** in the MTBC: any two strains differ by only around 2,500 SNPs over the whole genome. To achieve a good resolution of recent evolution (e.g. during an outbreak), whole genome sequences are thus extremely useful. With whole genome sequences, resolution can be further increased by not only considering SNPs, but also other types of mutations, in particular short insertions and deletions (as e.g. shown in [this study]( https://doi.org/10.1371/journal.ppat.1008357)).
>
> c) A large proportion of DNA polymorphisms in the MTBC are **singletons**, that is, variants present only in a single strain. This adds to the problem of low diversity, since singletons are not informative about tree topology.
>

> > d) In this workshop, and indeed in many studies of the MTBC, SNPs are called not against H37Rv, but against a reconstructed ancestral genome. This means that the number of SNPs identified does not reflect the evolutionary distance from some random strain like H37Rv, but from the most recent common ancestor of the MTBC. Below we will see that this has implications for the interpretation of a tree.

Reviewer 2: I always rather like Figure 3 from [this preprint](https://www.biorxiv.org/content/10.1101/403824v1.full) for this point.

Author: Indeed! I added a reference to the figure in the preprint

>
{: .comment}

# The alignment
Aligned DNA or protein sequences are the starting material for phylogenetic inference with molecular data. Here we will make use of the single nucleotide polymorphisms (SNPs) you obtained in the previous tutorials on SNP calling and transmission clusters, reflecting the diversity of 19 MTBC strains and 1 strain of **M. canettii**. The latter is included to have an outgroup, allowing us to root the phylogeny.

> Recall that the alignment here was generated using SNPs called from reference-aligned short reads. A frequently used alternative approach to obtain a phylogeny from short read data is to a) assemble the genomes (see the numerous [Galaxy tutorials](https://training.galaxyproject.org/training-material/topics/assembly) on this topic), b) annotate genes, c) extract genes present in all strains (the "core" genes), d) align the core genes. This approach underlies core genome multilocus sequence typing (cgMLST), which is often used to genotype bacterial pathogens (e.g. [Zhou et al. 2021](https://doi.org/10.1093/bioinformatics/btab234)).

Author: I'm changing this part, putting the alternative approach in a comment box. A full day of the workshop is about mapping and variant calling, so I thought I can keep it short here. Regarding the 'why', I'm actually not so sure (I worked on plants before). I'd expect to get similar phylogenetic resolution from SNP and core gene alignments, no?

> Recall that the alignment here was generated using SNPs called from reference-aligned short reads. A frequently used alternative approach to obtain a phylogeny from short read data is to a) assemble the genomes (see the numerous [Galaxy tutorials](https://training.galaxyproject.org/training-material/topics/assembly) on this topic), b) annotate genes, c) extract genes present in all strains (the "core" genes), d) align the core genes. This approach underlies core genome multilocus sequence typing (cgMLST), which is often used to genotype bacterial pathogens (e.g. [Zhou et al. 2021](https://doi.org/10.1093/bioinformatics/btab234)).

Reviewer 2: The 'why' is related to the relative lack of diversity in _M. tuberculosis_ and friends. For other bacteria there is a lot more recombination going on and, of course, horizontal gene transfer. So you don't have as simple a reference to map to and call SNPs off. So de-novo assembly followed by looking for a core genome and then building a phylogeny from the core genome is often used.

Having said that, there are limitations to using the H37Rv genome (that aren't addressed by using the inferred ancestral reference)... I talk a bit about that [here](https://youtu.be//yCC9EmLuyJg). It doesn't impact phylogeny *that much* though, as I understand it.

> Recall that the alignment here was generated using SNPs called from reference-aligned short reads. A frequently used alternative approach to obtain a phylogeny from short read data is to a) assemble the genomes (see the numerous [Galaxy tutorials](https://training.galaxyproject.org/training-material/topics/assembly) on this topic), b) annotate genes, c) extract genes present in all strains (the "core" genes), d) align the core genes. This approach underlies core genome multilocus sequence typing (cgMLST), which is often used to genotype bacterial pathogens (e.g. [Zhou et al. 2021](https://doi.org/10.1093/bioinformatics/btab234)).

Author: Thanks, makes sense! I guess assemblies are also trickier to work with than aligned reads, problems more difficult to detect, so given a choice between the two...

An alignment of SNPs looks something like this. Each row is a different strain, each column a position in the reference genome. Because this alignment is based on SNPs, it contains only variable positions. Inspecting the alignment is always a good idea, because also in phylogenetics the principle of 'garbage in, garbage out' applies.

<img src="./images/MEGA_alignment.png" alt="drawing" width="500"/>

> <br/><br/>

Reviewer: ```suggestion
![drawing](./images/MEGA_alignment.png){:width="500"}

Author: Why this alternative image formatting here and not for the other images?

Reviewer: No, exactly, please do it for all images! I just didn't feel like making the changes for all of them, so this is just an example, please replicate it for the others :)

>
>    ```
>    https://zenodo.org/record/6010176/SNP_alignment.fasta
>    ```
>
>    {% snippet faqs/galaxy/datasets_import_via_link.md %}
>
> 3. Take a look at the alignment
>
>   Once the file is loaded, select it in the history on the right. Click on the 'Visualize' icon and and try both, the 'Editor' and the 'Multiple Sequence Alignment'. Can you see one of the MTBC particularities mentioned above, the predominance of singletons? Also take a look at the M. canetti sequence: being the outgroup, it has a large number of SNPs.
>
{: .hands_on}

# Estimate a phylogeny
There are numerous methods to infer phylogenetic trees, but the most frequently used with large-scale molecular data are based on maximum likelihood and Bayesian inference. The details of how these methods construct trees from an alignment are beyond the scope of this introductory course. To be able to read trees, it is not necessary to know the statistical and computational details of how the trees are estimated. The books listed in the [Resources](#Resources) section provide in-depth introductions into the different principles of phylogenetic inference, in particular Baum & Smith 2013 and Yang 2014.

> In this tutorial, we will use the maximum likelihood method [RAxML](https://cme.h-its.org/exelixis/web/software/raxml/) to estimate a phylogenetic tree for the 20 strains.

Author: Hm, RAxML-ng is not in the Galaxy toolshed, while I've missed IQ-TREE. I'll take a look and possibly change to IQ-TREE

> In this tutorial, we will use the maximum likelihood method [RAxML](https://cme.h-its.org/exelixis/web/software/raxml/) to estimate a phylogenetic tree for the 20 strains.

Reviewer 2: Yep, someone needs to write a RAxML-ng wrapper. *sigh* A nice topic for a future hackathon?

> In this tutorial, we will use the maximum likelihood method [RAxML](https://cme.h-its.org/exelixis/web/software/raxml/) to estimate a phylogenetic tree for the 20 strains.

Author: I'll stick to RAxML for the course. Given the introductory level and the focus on reading rather than estimating trees, it doesn't really make a difference, no?

>
> 1. {% tool [Phyogenetic reconstruction with RAxML](toolshed.g2.bx.psu.edu/repos/iuc/raxml/raxml/8.2.4+galaxy2) %} with the following parameters:
>    - {% icon param-file %} *"Source file with aligned sequences"*: `output` (Input dataset)
>    - *"Model type"*: `Nucleotide`
>    - *"RAxML options to use"*: `Required options only`
>
> 2. The RAxML output we are interested in is the "Best-scoring ML tree". Select it in you Galaxy history and take a look at it with the different visualization options offered by Galaxy.
>
{: .hands_on}

# Visualize and manipulate the tree
Phylogenetic trees are great tools because they are at the same time quantitative (we can do calculations on branch lengths, estimate uncertainty of a tree topology etc.) and visually appealing, allowing to actually "see" biologically interesting patterns. Often this requires some tweaking of the tree, for example by coloring parts of the tree according to some background information we have about the samples. Here, to obtain an overview of the different strains present in the country (given the limitations of the sampling), our goal is to create a rooted tree in which colors indicate different MTB lineages.

> We will use R to plot and manipulate the phylogeny obtained from RAxML. The code to produce the figures is shown in the boxes below. You can execute it by starting RStudio within Galaxy, as explained [here](https://shiltemann.github.io/training-material/topics/galaxy-interface/tutorials/rstudio/tutorial.html). This is not required to finish this tutorial, but if you have used R before, it might be worthwile to go through the code, modify it, and explore the numerous phylogenetics packages and functions in R.

Author: changed

> We will use R to plot and manipulate the phylogeny obtained from RAxML. The code to produce the figures is shown in the boxes below. You can execute it by starting RStudio within Galaxy, as explained [here](https://shiltemann.github.io/training-material/topics/galaxy-interface/tutorials/rstudio/tutorial.html). This is not required to finish this tutorial, but if you have used R before, it might be worthwile to go through the code, modify it, and explore the numerous phylogenetics packages and functions in R.

Author: Yes, I was also experimenting with iTOL, but I found it difficult to come up with an exercise which would not consist in boring clicking around...

> >
> {: .solution}
>
{: .question}

## Root the tree
To make the phylogeny better interpretable, will now root it and add some additional information. First, we root the tree and then exclude the canettii strain, such that patterns within the MTBC become more clear. This already looks better, the tree topology stands out more clearly now, and we can identify groups of closely related strains.

```{r}
# Root the tree
tree_rooted <- root(tree, "ERR313115.fastq.vcf")

# Remove the outgroup to make distances within MTB more clear
tree_rooted <- drop.tip(tree_rooted, "ERR313115.fastq.vcf")
tree_rooted$root.edge <- 0.005
plot(tree_rooted, root.edge = T, cex=0.6)
```

<img src="./images/tree_rooted.svg" alt="drawing" width="600"/>

## Show the different lineages present in the sample
A first piece of information we now want to add to the phylogeny is to which lineage the strains belong. This will allow us to assess whether our tree is consistent with the known phylogeny of the MTBC, shown in Figure 1A, and to visualize which lineages are present in our samples. The information to which lineage a strain belongs can be found in the output of TB-profiler.

> ```{r}

Reviewer: ```suggestion
```r
```

Author: changed in all 3 instances

>
> > ### {% icon solution %} Solution
> >
> > The VCF files hold the key to explain our puzzling observation: ERR1203059.vcf contains not a single SNP, ERR5987300.vcf only 81 SNPs. By contrast, the other strains have between 750 and 1250 SNPs. What happened here?
> >
> {: .solution}
>
{: .question}

> ### {% icon question %} Question 3
>
> Recall the clusters identified in the previous tutorial, reproduced below. How do these clusters show up in the phylogenetic tree? What additional information does the tree contain?
>
| Sample       | Cluster_id | DR profile | Clustering  |
|--------------|------------|------------|-------------|
| ERR5987352   | 10         | Pre-MDR    | Clustered   |
| ERR6362484   | 10         | Pre-MDR    | Clustered   |
| ERR6362138   | 12         | MDR        | Clustered   |
| ERR6362156   | 12         | Pre-XDR    | Clustered   |

> | ERR6362253   | 12         | MDR        | Clustered   |

Reviewer: this table seems to be lacking the blockquote indentation, which will break the box

Author: blockquotes added

---
layout: tutorial_hands_on
title: Tree thinking for tuberculosis evolution and epidemiology
zenodo_link: 'https://zenodo.org/record/6010176'
tags:
  - prokaryotes
  - phylogenetics
  - evolution

> level: Beginner

Reviewer: ```suggestion
level: Introductory
```

> objectives:

Reviewer: ```suggestion
questions:
- How do I make a phylogeny?
- How can I visualise and interpret the tree?
objectives:
```

Author: This was adapted.

An alignment of SNPs looks something like this. Each row is a different strain, each column a position in the reference genome. Because this alignment is based on SNPs, it contains only variable positions. Inspecting the alignment is always a good idea, because also in phylogenetics the principle of 'garbage in, garbage out' applies.

![Alignment](./images/MEGA_alignment.png){:width="500"}

## Get the data
> ### {% icon hands_on %} Hands-on: Obtain your data
>
> 1. Make sure you have an empty analysis history. Give it a name.
>
>    {% snippet faqs/galaxy/histories_create_new.md %}
>
> 2. Import the following files from [Zenodo](https://zenodo.org/record/6010176) or from the shared data library
>
>    ```

> >    https://zenodo.org/record/6010176/SNP_alignment.fasta

Reviewer 2: Should be https://zenodo.org/record/6010176/files/SNP_alignment.fasta

Author: Changed, thanks!
